File: Algorithm/python_source/charpter17_greedy/level3/can_jump.py
# coding=utf-8

class CanJump:
   def canJump(self, nums) :
        n, rightmost = len(nums), 0
        for i in range(n):
            if i <= rightmost:
                rightmost = max(rightmost, i + nums[i])
                if rightmost >= n - 1:
                    return True
        return False


if __name__ == '__main__':
    ratings = [1, 0, 2]
    canJump = CanJump()
    print canJump.canJump(ratings)

File: Algorithm/c_or_cpp/charpter17_greedy/level3/can_jump.cpp
#include <iostream>
#include <vector>

using namespace std;

bool canJump(vector<int>& nums) {
    int n = nums.size();
    int rightmost = 0;
    for (int i = 0; i < n; ++i) {
        if (i <= rightmost) {
            rightmost = max(rightmost, i + nums[i]);
            if (rightmost >= n - 1) {
                return true;
            }
        }
    }
    return false;
}

int main() {
    vector<int> nums = {3, 1, 4, 8, 2, 7};
    bool result = canJump(nums);
    cout << "Can the player jump from " << nums.back() << " to " << nums.back() + 7 << "? " << (result ? "Yes" : "No") << endl;
    return 0;
}